lib/AuthStateManager.ts

```diff
@@ -0,0 +1,129 @@
+import { AuthSdkError } from './errors';
+import PromiseQueue from './PromiseQueue';
+import { omit } from './util';
+import { 
+  ACCESS_TOKEN_STORAGE_KEY, 
+  ID_TOKEN_STORAGE_KEY 
+} from './constants';
+import { 
+  AuthState,
+  Tokens, 
+  AccessToken, 
+  IDToken,
+  UpdateAuthStateOptions
+} from './types';
+import { OktaAuth } from './browser';
+
+const DEFAULT_AUTH_STATE = { 
+  isPending: true,
+  isAuthenticated: false,
+  idToken: null,
+  accessToken: null,
+};
+const EVENT_AUTH_STATE_CHANGE = 'authStateChange';
+
+class AuthStateManager {
+  private sdk: OktaAuth;
+  private pending: { tokens: Tokens };
+  private authStateQueue: PromiseQueue;
+  private authState: AuthState;
+
+  constructor(sdk: OktaAuth) {
+    if (!sdk.emitter) {
+      throw new AuthSdkError('Emitter should be initialized before AuthStateManager');
+    }
+
+    this.sdk = sdk;
+    this.pending = { tokens: {} };
+    this.authStateQueue = new PromiseQueue();
+    this.authState = { ...DEFAULT_AUTH_STATE };
+
+    // Listen on tokenManager events to sync update tokens in memory (this.pending)
+    // And push async update authState and emit process into PromiseQ
+
+    // "added" event is emitted in both add and renew process
+    // Only listen on "added" (instead of both "added" and "renewed") to limit authState re-evaluation
+    sdk.tokenManager.on('added', (key, token) => {
+      this.pending.tokens = { ...this.pending.tokens, [key]: token };
+      this.authStateQueue.push(
+        this.updateAuthState, 
+        this,
+        { ...this.pending.tokens }
+      );
+    });
+    sdk.tokenManager.on('removed', (key) => {
+      this.pending.tokens = omit(this.pending.tokens, key);
+      this.authStateQueue.push(
+        this.updateAuthState, 
+        this,
+        { ...this.pending.tokens }
+      );
+    });
+  }
+
+  getAuthState(): AuthState {
+    return this.authState;
+  }
+
+  async initialAuthState(): Promise<void> {
+    let accessToken = await this.sdk.tokenManager.get(ACCESS_TOKEN_STORAGE_KEY) as AccessToken;
+    if (accessToken && this.sdk.tokenManager.hasExpired(accessToken)) {
+      accessToken = null;
+    }
+    let idToken = await this.sdk.tokenManager.get(ID_TOKEN_STORAGE_KEY) as IDToken;
+    if (idToken && this.sdk.tokenManager.hasExpired(idToken)) {
+      idToken = null;
+    }
+    this.updateAuthState({ accessToken, idToken });
+  }
+
+  updateAuthState({
+    accessToken, 
```

Comment:
why are the tokens being passed to the function instead of being read from the storage?

Reply:
As the listeners from TokenManager are able to provide both keys and tokens, pass tokens from listeners can reduce unnecessary async calls. https://github.com/okta/okta-auth-js/pull/453/files#diff-c87977be240b146c8cc1ab3c11a6cc66R48

Also, as we want to provide sequential data flow, it makes sense to bind tokens to specific event triggered func in the promiseQ.

Reply:
The first point (unnecessary async calls) is an optimization. I would like to avoid making optimizations in this method, these async calls are not expensive.

The second point (sequential data flow, bind to tokens), has me thinking. I think this highlights a problem with the promiseQ approach. If updateAuthState runs in the future it may have inconsistent state. I actually think your original approach of cancelling if an updateState is in progress would be more appropriate. 

updateState takes some time to run. If another request to updateState comes in while an update is already in progress, we should cancel/discard the in progress operation and start a fresh evaluation.

I understand that this contradicts the other approach of sequential fine-grained state updates, but I think it will lead to the most consistent output. There is no way we can pre-bind all possible variables to `isAuthenticated` or other aspects of the update cycle. These will always be done using **current** values. If a new call to `updateAuthState` comes in while we are running the cycle it indicates **something** has been updated, but we don't know what it is. (I would say we don't WANT to know what it is, this escalates the complexity). But we can say that our current cycle is now dirty/invalid, cancel it, and start a fresh cycle from the beginning.

So I think it's best if `updateAuthState` takes no parameters and returns no value. But it can cancel and restart itself if a concurrent request comes in.

Reply:
I was actually convinced by the sequential data flow, as cancelable data flow DO eat up states.
And for the `AuthStateManager`, it's designed to be driven by tokens updates from TokenManager, as we have scoped the variables (tokens), it probably just fine to only bind the tokens when update.

Also, the async call may not cheap, right now I think the only reason we want to keep the `tokenManager.get` call async is because we want to extend it to handle network request in the future (not cheap). Per current needs, we don't need the method to be async.

Reply:
If they use, say IndexedDB for their token storage, that's async.

If they use any kind of async access to their own app storage, that's async.

Reply:
Maybe it just make sense to keep `updateAuthState` private? 
then public APIs can be
`initialAuthState` to provide initial the in-memory authState
`onAuthStateChange` and `offAuthStateChange` to add/remove listeners for authState change.
`getAuthState` to synchronously get latest evaluated authState

Reply:
The developer could change tokens out from under us - updateAuthState() is how they let the system know they've done that.

Reply:
I am thinking probably it's better to let devs to emit a token event than call updateAuthState directly? I like the idea of maintaining a sequential data flow by tokens events, but call the method directly looks like it would interrupt the flow directly by bypassing the event step, as the process may not leverage current promiseQ to keep the data flow in sequence. 

imo, the data flow (sequential, if we can manage it proper I still prefer this way) can always be a top-down process:
Default: TokenManager emit an event -> AuthStateManager react to the event by pushing updateState into promiseQ -> authStateChange event be emitted after evaluation
Customized (manage token changes directly): Emit event by using method from tokenManager (may need to provide public API to emit same events) -> AuthStateManager take over the rest process
Opt-out: the devs manager authState update by themselves.

Reply:
I'm not really thrilled at the idea of devs using the emitter to emit their own events - that puts us into a weird middleware state - but since they control the token storage that might be the right answer...but I'm not excited about it.

This would definitely involve a new section about managing the token if we go that way.

Reply:
I'm not sure we would need to document `updateAuthState` unless the users needs to call it. Why would they need to call it? What if their `isAuthenticated` function uses information we don't know about. When that information is updated, they could call `updateAuthState` to cause a re-evaluation of their own `isAuthenticated` function.

Reply:
...Did you just say you don't think we need to document the function, then list an explicit scenario where they would need to call it?

Reply:
But that's a good point - isAuthenticated() being overridden to contain things other than tokens means asking them to emit doesn't work, we'd need them to call `updateAuthState()`

Reply:
@aarongranick-okta I think there might be scenarios like users want to manager tokens under keys other than what we used in the SDK. Then they would want to customized the flow.
@swiftone I think we can provide methods in TokenManager with emit bound instead of letting users touch the `emitter` directly.

Reply:
@shuowu I don't understand the benefit of hiding `updateAuthState` behind an event. Why is this is a better interface than calling the method directly?

Reply:
Before the concern is by using authStateManager.updateAuthState directly, the async process cannot be added to the promiseQ, which may mess up the sequential data flow. So I mentioned an option to make all updates event-driven. But looks like you guys are not very into that approach.

The way I just added is to expose `oktaAuth.updateAuthState` without any args needed. This method can put the async into the promiseQ, which keep the data flow still in good state. For the implementation details, please see https://github.com/okta/okta-auth-js/pull/453/files#diff-75082c639aefcc47b3b8df57c8870867R331